gax.go

```diff
@@ -1,191 +1,169 @@
-		length := streamable.Len()
-		for i := 0; i < length; i++ {
-			if !iter(streamable.GetData(i)) {
-				return nil
-			}
-		}
-		if err := streamable.NextPage(); err == io.EOF {
-			return nil
-		} else if err != nil {
-			return err
-		}
+func (opts callOptions) Resolve(s *CallSettings) *CallSettings {
+	for _, opt := range opts {
+		opt.Resolve(s)
 	}
+	return s
 }
 
-// Head returns the first element from the page stream.
-func Head(ctx context.Context, streamable PageStreamable, opts ...CallOption) (result interface{}, err error) {
-	called := false
-	err = PageStream(ctx, streamable, func(element interface{}) bool {
-		called = true
-		result = element
-		return false
-	}, opts...)
-	if !called {
-		return nil, errors.New("PageStreamable is empty.")
-	}
-	return result, err
+// Encapsulates the call settings for a particular API call.
+type CallSettings struct {
+	// The client-side timeout for API calls.
+	Timeout time.Duration
+	// Determines whether and how to retry on transient errors.
+	RetrySettings RetrySettings
 }
 
-// Take returns the first |count| elements in the page stream.
-func Take(ctx context.Context, streamable PageStreamable, count int, opts ...CallOption) ([]interface{}, error) {
-	if count <= 0 {
-		return nil, nil
-	}
-	result := make([]interface{}, 0, count)
-	i := 0
-	err := PageStream(ctx, streamable, func(element interface{}) bool {
-		result = append(result, element)
-		i++
-		return i < count
-	}, opts...)
-	return result, err
-}
-
-// ToArray converts the page stream into a single array.
-func ToArray(ctx context.Context, streamable PageStreamable, opts ...CallOption) ([]interface{}, error) {
-	var result []interface{}
-	err := PageStream(ctx, streamable, func(element interface{}) bool {
-		result = append(result, element)
-		return true
-	}, opts...)
-	return result, err
+func (s *CallSettings) Resolve(c *CallSettings) {
+	c.Timeout = s.Timeout
+	c.RetrySettings = s.RetrySettings
+}
+
+// Per-call configurable settings for retrying upon transient failure.
+type RetrySettings struct {
+	// A list of Google API canonical error codes upon which a retry should
+	// be attempted. If nil, the call will not retry.
+	RetryCodes []string
+	// Configures the retry exponential backoff algorithm.
+	BackoffSettings BackoffSettings
+}
+
+// Parameters to the exponential backoff algorithm for retrying.
+type BackoffSettings struct {
+	// The initial delay time, in milliseconds, between the completion of
+	// the first failed request and the initiation of the first retrying
+	// request.
+	InitialRetryDelayMillis time.Duration
+	// The multiplier by which to increase the delay time between the
+	// completion of failed requests, and the initiation of the subsequent
+	// retrying request.
+	RetryDelayMultiplier float64
+	// The maximum delay time, in milliseconds, between requests. When this
+	// value is reached, `RetryDelayMultiplier` will no longer be used to
+	// increase delay time.
+	MaxRetryDelayMillis time.Duration
```

Comment:
Are there specific reasons to flatten the settings?  I personally like previous design of defining initial/max/multiplier into a single struct and defining methods (well, I like this mostly because I designed it).

I don't think there's a strict reason to follow the exact structure of the setting structs, and in Go context, I believe splitting into a struct is clearer.